microservices-connector/cmd/router/main.go

```diff
@@ -727,10 +731,155 @@ func handleMultipartError(writer *multipart.Writer, err error) {
 	log.Error(err, "Error during multipart creation")
 }
 
+// create a handler to handle traffic to /ui
+// if the payload is empty, redirect to ui service
+// if there's payload, format the payload and redirect to backend service
+func mcUiHandler(w http.ResponseWriter, req *http.Request) {
+	// redirect traffic to ui pod if payload is empty
+	// redirect traffic to mcGraphHandler if payload is not empty
+	defaultNode := mcGraph.Spec.Nodes[defaultNodeName]
+	for i := range defaultNode.Steps {
+		step := &defaultNode.Steps[i]
+		if UI == step.StepName {
+			body, err := io.ReadAll(req.Body)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+
+			// if no payload included in the request, redirect request to UI
+			if len(body) == 0 {
+				serviceURL := getServiceURLByStepTarget(step, mcGraph.Namespace)
+				targetURL, err := url.Parse(serviceURL)
+				if err != nil {
+					http.Error(w, err.Error(), http.StatusInternalServerError)
+					return
+				}
+				proxy := httputil.NewSingleHostReverseProxy(targetURL)
+				proxy.ServeHTTP(w, req)
+			} else {
+				// if payload exists, format payload and redirect to backend services
+				var data map[string]interface{}
+				parsedData := map[string]string{}
+
+				// find the first hop for the pipeline
+				var nextHop *mcv1alpha3.Step
+				for i := range defaultNode.Steps {
+					nextHop = &defaultNode.Steps[i]
+					if nextHop.InternalService.IsDownstreamService {
+						// skip downstream service
+						continue
+					}
+					break
+				}
+
+				// set the corresponding keyword for input data
+				var key string
+				switch nextHop.StepName {
+				case "Embedding":
```

Comment:
Hi @Ruoyu-y, I think there should be only 2 cases: request from the mega service and directly to embedding service. Am I right?

Reply:
Let me expain, since GMC wants to pass through the payload directly to the pipeline. The format of the payload must fit the requirement of the pipeline's first component. I found that in the examples, the first hop of the pipeline is either embedding or llm. So here, i am trying to find the first hop of the pipeline and use the corresponding keyword to format the payload.
I am not sure why there would be request from the mega service in the GMC case, can you explain?